[Lim Jin Hao] iP

JinHao-L opened the pull request.

Commit:
Add Gradle support

JinHao-L committed:
Add greet, echo, exit functionality

JinHao-L committed:
Implement add, list functionality

JinHao-L committed:
Add ability to mark tasks as done

JinHao-L committed:
Add tasks type: todos, events, deadlines

JinHao-L committed:
Add automated text UI testing

JinHao-L committed:
Handle errors

JinHao-L committed:
Add support for deleting tasks from list

JinHao-L committed:
Add enum for allowed commands

JinHao-L committed:
Rename variables to comply with naming convention

JinHao-L committed:
Add save and load feature

JinHao-L committed:
Add date and time support for events and deadlines

JinHao-L committed:
Merge branch 'branch-Level-7' into master

JinHao-L committed:
Merge branch 'branch-Level-8' into master

# Conflicts:
#	src/main/java/Deadline.java

JinHao-L committed:
Add clear tasks command for testing consistency

JinHao-L committed:
Use more OOP

JinHao-L committed:
Organise classes into packages

JinHao-L committed:
Add JUnit Tests

JinHao-L committed:
Package the app as a JAR file

JinHao-L committed:
Add javadoc comments

JinHao-L committed:
Add find command to search for task using keywords

JinHao-L committed:
Tweak the code to comply with coding standard

JinHao-L committed:
Merge branch 'branch-A-JavaDoc' into master

JinHao-L committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/alice/Alice.java
#	src/main/java/alice/command/DeadlineCommand.java
#	src/main/java/alice/parser/Parser.java

JinHao-L committed:
Merge branch 'branch-Level-9' into master

# Conflicts:
#	src/main/java/alice/parser/Parser.java
#	src/main/java/alice/task/Task.java
#	src/main/java/alice/task/TaskList.java

JinHao-L committed:
Add JavaDocs for FindCommand

JinHao-L committed:
Merge remote-tracking branch 'origin/add-gradle-support' into master

JinHao-L committed:
Automate project builds using Gradle

JinHao-L committed:
Use checkStyle to detect coding style violations

JinHao-L committed:
Use var-args

JinHao-L committed:
Set up Java FX

JinHao-L committed:
Build basic layout for GUI

Door-oof commented on src/main/java/alice/Alice.java:
```diff
@@ -0,0 +1,69 @@
+
+    /**
+     * Creates an ALICE program that reads data from specified location.
+     * If the file does not exist, creates the appropriate directory and file
```
There should be a punctuation behind the description

JinHao-L replied:
Yup I have updated it on my end. Thanks for the input.

Door-oof commented on src/main/java/alice/command/InvalidCommandException.java:
```diff
@@ -0,0 +1,18 @@
+import alice.AliceException;
+
+/**
+ * Thrown when the command given by the user is invalid.
```
Perhaps rephrasing the sentence so that the action word "Throws" is used instead of "Thrown"?

JinHao-L replied:
Ooh, nice catch! Thanks for the input.

Door-oof commented on src/main/java/alice/task/Event.java:
```diff
@@ -0,0 +1,86 @@
+            .ofPattern("EEEE, MMM dd uuuu, ha");
+
+    private final LocalDateTime on;
+    private final boolean includesTime;
```
Perhaps a more intuitive variable name to sound more like a boolean?

JinHao-L replied:
Agree, changed it to hasTime instead.
Thanks for the input. 👍

Door-oof commented on src/main/java/alice/task/Deadline.java:
```diff
@@ -0,0 +1,86 @@
+            .ofPattern("EEEE, MMM dd uuuu");
+
+    private final LocalDateTime by;
+    private final boolean includesTime;
```
Perhaps a more intuitive variable name to sound more like a boolean?

JinHao-L replied:
Agree. Updated on my end. Thanks for the input.

Door-oof reviewed (commented):
Overall, I find your code readable and that it mostly adheres to the coding standards. There were a few small minor coding standard violations that were probably overlooked, but a nice job nonetheless.

Trav1sT commented on src/main/java/alice/command/ClearCommand.java:
```diff
@@ -0,0 +1,47 @@
+        String confirmation = ui.getFeedback("Are you sure you want to clear all tasks? [Y/N]");
+        ui.displayLine();
+
+        if (confirmation.trim().toLowerCase().startsWith("y")) {
```
Perhaps `.strip()` can be used here instead as it is unicode compliant

JinHao-L replied:
Oh, I didn't know that. I will read up on it and change it accordingly.
Thanks for the suggestion! 👍

Trav1sT commented on src/main/java/alice/command/HelpCommand.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public void process(TaskList tasks, Ui ui, StorageFile storageFile) {
+        String commandFormat = "    %-35s | %s\n";
+
+        StringBuilder output = new StringBuilder("These are the commands in my dictionary:\n");
+        output.append(String.format(commandFormat, ListCommand.USE_CASE, ListCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, FindCommand.USE_CASE, FindCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, ClearCommand.USE_CASE, ClearCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, DoneCommand.USE_CASE, DoneCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, DeleteCommand.USE_CASE, DeleteCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, TodoCommand.USE_CASE, TodoCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, DeadlineCommand.USE_CASE, DeadlineCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, EventCommand.USE_CASE, EventCommand.DESCRIPTION));
+        output.append(String.format(commandFormat, USE_CASE, DESCRIPTION));
+        output.append(String.format(commandFormat, ByeCommand.USE_CASE, ByeCommand.DESCRIPTION));
+
```
I like how 'String.format()' is used and StringBuilder is used as it is not a short string.

JinHao-L replied:
Thank you! :D

Trav1sT commented on src/main/java/alice/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+     */
+    private static List<DateTimeFormatter> createDateFormats() {
+        // List of acceptable date time format with optional time/year
+        List<String> knownPatterns = Arrays.asList(
+                "d/M[/uuuu][ HHmm]", "d-M[-uuuu][ HHmm]",
+                "M/d[/uuuu][ HHmm]", "M-d[-uuuu][ HHmm]",
+                "uuuu/M/d[ HHmm]", "uuuu-M-d[ HHmm]",
+                "d-MMM[-uuuu][ HHmm]", "d MMM[ HHmm]"
+        );
+
```
I like how a list of patterns is used to allow more flexibility in user's input for date and time.

JinHao-L replied:
Thank you!

Trav1sT commented on src/main/java/alice/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+            argument = "";
+        }
+
+        if (ListCommand.hasCommandWord(cmd)) {
+            // Command to display task list
+            return new ListCommand();
+        } else if (FindCommand.hasCommandWord(cmd)) {
+            // Command to find specific tasks
+            return parseFindKeywords(argument);
+        } else if (ClearCommand.hasCommandWord(cmd)) {
+            // Command to clear task list
+            return new ClearCommand();
+        } else if (HelpCommand.hasCommandWord(cmd)) {
+            // Command to get list of commands
+            return new HelpCommand();
+        } else if (DoneCommand.hasCommandWord(cmd)) {
+            // Command to mark task as done
+            return parseDoneInput(argument);
+        } else if (DeleteCommand.hasCommandWord(cmd)) {
+            // Command to mark task as done
+            return parseDeleteInput(argument);
+        } else if (TodoCommand.hasCommandWord(cmd)) {
+            // Command to add to-do
+            return parseTodoInput(argument);
+        } else if (DeadlineCommand.hasCommandWord(cmd)) {
+            // Command to add deadline
+            return parseDeadlineInput(argument);
+        } else if (EventCommand.hasCommandWord(cmd)) {
+            // Command to add event
+            return parseEventInput(argument);
+        } else if (ByeCommand.hasCommandWord(cmd)) {
+            return new ByeCommand();
+        } else {
+            // Invalid command
+            throw new InvalidCommandException("Sorry I cannot register that command!\n"
+                    + "Use 'help' command to see the lists of available command");
+        }
```
I feel like this can be shortened with a loop or switch case. Did you do it this way because it might be more readable?

JinHao-L replied:
Thanks for raising this issue.

I wanted to use a switch case for this actually, however, I am not sure how to go about it as I am using a static method of different class as my condition. As for loops, I'm not sure how to go about looping through different classes. :(

Do let me know if you have any suggestion on how I should approach this. 👍

Trav1sT commented on src/main/java/alice/parser/Parser.java:
```diff
@@ -0,0 +1,251 @@
+            // Empty description
+            throw new InvalidCommandException("The deadline description cannot be left empty.");
+        } else if (argument.endsWith("/by")) {
+            // Empty date
+            throw new InvalidCommandException("You cannot create an deadline without the date.");
+        } else {
+            // No /by marker
+            throw new InvalidCommandException("I can't find the deadline date. "
+                    + "Did you forget to add '/by'?");
+        }
+    }
+
+    /**
+     * Parses the details given by the user for creating a new event.
+     *
+     * @param argument the event details input given by user.
+     * @return the <code>EventCommand</code> with the indicated details.
+     * @throws InvalidCommandException if the user gives an invalid description and/or datetime.
+     */
+    private static EventCommand parseEventInput(String argument) throws InvalidCommandException {
+        String[] arguments = argument.split(" /on ", 2);
+        if (arguments.length == 2 && !arguments[1].isBlank()) {
+            String description = arguments[0];
+            String dateTime = arguments[1];
+
+            LocalDateTime eventDateTime = parseDateTime(dateTime);
+            return new EventCommand(description, eventDateTime);
+        } else if (argument.isBlank()) {
+            // Empty event description
+            throw new InvalidCommandException("The event description cannot be left empty.");
+        } else if (argument.endsWith("/on")) {
+            // Empty start-end time
+            throw new InvalidCommandException("You cannot create an event without a date/time.");
+        } else {
+            // No /on marker
+            throw new InvalidCommandException("I can't find the date/time of the event. "
+                    + "Did you forget to add '/on'?");
+        }
+    }
+
```
To me, one function for each type of command seems a bit excessive. Perhaps, there are other ways you would like to shorten this and make it seem less repetitive?

JinHao-L replied:
Thanks for the input! I agree with what you mention.

I have 2 ideas in mind at the moment. I could either shift each part to their respective command class under a parseArgument function or create a factory method for each command class that parses the argument if needed.

I'll look into it and see which would be more suitable. 
Do let me know if you have any suggestions 👍

Trav1sT commented on src/test/java/alice/parser/ParserTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import alice.command.InvalidCommandException;
+
+public class ParserTest {
+    @Test
+    void parseDateTimeTest() throws InvalidCommandException {
+        LocalDateTime dateTime = Parser.parseDateTime("12-Aug 1350");
+        assertEquals(LocalDateTime.of(2020, 8, 12, 13, 50), dateTime);
+    }
+}
```
Given the multiple formats allow, perhaps you would want to test other valid and invalid inputs as well.

JinHao-L replied:
The JUnit test is still an ongoing work and due to time constraint, I did not include a lot of tests. 

I will most likely add on this test in the future updates. Thanks for the input! 👍

Trav1sT reviewed (commented):
LGTM. I like how the comments are used generously and how they are phrased which allows me to understand what to expect of the function easily. Structure of the code seems good.

JinHao-L committed:
Add GUI for Alice

JinHao-L committed:
Add JavaDoc

JinHao-L committed:
Merge branch 'branch-Level-10' into master

JinHao-L committed:
Update GUI for bottom input bar

JinHao-L committed:
Add JUnit tests. Improve find implementation.

Add Junit tests for testing tasks related methods. Improve find implementation to search by whole words rather than
substring.

JinHao-L committed:
Improve code quality

JinHao-L committed:
Shift parse logic into respective commands

JinHao-L committed:
Add assertions

JinHao-L committed:
Use streams

JinHao-L committed:
Create CommandType to reduce repetition in parser

JinHao-L committed:
Merge pull request #1 from JinHao-L/branch-A-CodeQuality

Improve code quality

JinHao-L committed:
Merge branch 'master' into branch-A-Lambdas

# Conflicts:
#	src/main/java/alice/util/Parser.java

JinHao-L committed:
Resolve merge conflicts and improve interface.

JinHao-L committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/alice/storage/StorageFile.java
#	src/main/java/alice/task/Todo.java

JinHao-L committed:
Merge pull request #2 from JinHao-L/branch-A-Lambdas

Lambdas

JinHao-L committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/alice/task/Task.java
#	src/main/java/alice/util/Parser.java

JinHao-L committed:
Merge pull request #3 from JinHao-L/branch-A-Assertions

Add assertions

JinHao-L committed:
Use Github Actions to setup Continuous Integration

JinHao-L committed:
Merge branch 'branch-A-CI' into master

JinHao-L committed:
Add separate parser for date and time

JinHao-L committed:
Add support for date and time in natural language.

Separately parses the date and time based on dictionary or known
formats.

JinHao-L committed:
Improve JUnit testing for date time parsing

JinHao-L committed:
Merge pull request #4 from JinHao-L/branch-C-NaturalDates

Branch C natural dates

JinHao-L committed:
Add background img to GUI

JinHao-L committed:
Improve GUI

JinHao-L committed:
Improve clear functionality and help output

JinHao-L committed:
Add a User Guide

JinHao-L committed:
Set theme jekyll-theme-architect

JinHao-L committed:
Set theme jekyll-theme-slate

JinHao-L committed:
Update User Guide

JinHao-L committed:
Merge branch 'master' of https://github.com/JinHao-L/ip into master

JinHao-L committed:
Update User Guide

JinHao-L committed:
Update User Guide

JinHao-L committed:
Remove setting where GUI opens in minimised form

JinHao-L committed:
Update User Guide

JinHao-L committed:
Update user guide

JinHao-L committed:
Add checks to prevent invalid string in desc

JinHao-L committed:
Update README.md

JinHao-L committed:
Update DateParserTest.java
